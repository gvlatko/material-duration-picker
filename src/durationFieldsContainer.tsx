import {Grid, GridProps} from '@mui/material';
import * as React from 'react'
import {ComponentType, Dispatch} from 'react'
import {DurationField, DurationFieldProps} from "./durationField";
import {DurationType, DurationView, Labels} from "./types";
import DefaultLabels from "./defaultLabelsEn.json";
import {durationToTime, timeToDuration} from "./utils";

export type DurationFieldsContainerProps = {
  views: DurationView[];
  value: number | undefined;
  setValue: Dispatch<number | undefined>
  labels?: Labels

  GridContainerProps?: GridProps;
  GridItemProps?: GridProps;
  DurationFieldProps?: Partial<DurationFieldProps>

  DurationFieldComp?: ComponentType<DurationFieldProps>;
}

export const DurationFieldsContainer = ({
  views = [],
  value,
  setValue,
  labels: _labels,
  GridItemProps,
  GridContainerProps,
  DurationFieldProps,
  DurationFieldComp = DurationField,
}: DurationFieldsContainerProps) => {
  const labels = {
    ...DefaultLabels,
    ..._labels
  }

  const duration = timeToDuration(value, views)

  return (
    <Grid container spacing={2} justifyContent='space-around' {...GridContainerProps}>
      {views.map((view, i) => (
        <Grid key={i} item xs {...GridItemProps}>
          <DurationFieldComp
            labels={labels}
            view={view}
            duration={duration}
            onConfirm={(v) => {
              setValue(
                durationToTime({
                  ...duration,
                  [view]: v
                })
              )
            }}
            {...DurationFieldProps}
          />
        </Grid>
      ))}
    </Grid>
  );
}

