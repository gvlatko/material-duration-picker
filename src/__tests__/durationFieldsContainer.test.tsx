import * as React from 'react'
import {render, waitFor, screen, fireEvent} from '@testing-library/react'
import {DurationFieldsContainer, DurationFieldsContainerProps} from "../durationFieldsContainer";
import {durationToTime, timeToDuration} from "../utils";
import {useState} from "react";

describe('DurationFieldsContainer', () => {
  const setValueMock = jest.fn()

  const Comp = ({
    views = ['hours', 'minutes'],
    duration = {hours: 5, minutes: 2},
    setValue= setValueMock,
    ...props
  }: any) => {
    return (
      <DurationFieldsContainer
        {...{
          views,
          value: durationToTime(duration),
          setValue,
          DurationFieldProps: {
            inputProps: {
              title: 'fieldTitle',
            },
            ...props.DurationFieldProps
          },
          ...props
        }}
      />
    )
  };

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(
        render(
          <Comp/>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  it('can render less views with greater duration', async () => {
    const utils = render(
      <Comp duration={{
        days: 1,
        hours: 1,
        minutes: 1,
        seconds: 30,
      }}/>
    );
    const fields: any[] = utils.getAllByTitle('fieldTitle')
    expect(fields[0].value).toBe("25")
    expect(fields[1].value).toBe("1.5")
  });

  it('can render less views with greater duration when smaller views are undefined', async () => {
    const utils = render(
      <Comp duration={{
        days: 2,
      }}/>
    );
    const fields: any[] = utils.getAllByTitle('fieldTitle')
    expect(fields[0].value).toBe("48")
    expect(fields[1].value).toBe("")
  });

  it('can setDuration to smaller than views show', async () => {
    const utils = render(
      <Comp duration={{
        hours: 1,
        minutes: 1
      }}/>
    );

    const fields: any[] = utils.getAllByTitle('fieldTitle')

    fireEvent.change(fields[0], {target: {value: '61'}})
    fireEvent.blur(fields[0])

    expect(fields[0].value).toBe('61')

    expect(setValueMock).toHaveBeenCalledWith(durationToTime({
      hours: 61,
      minutes: 1,
    }));
  })

  it('can setDuration to smaller than views show', async () => {
    const utils = render(
      <Comp duration={{
        hours: 1,
        minutes: 1
      }}/>
    );

    const fields: any[] = utils.getAllByTitle('fieldTitle')

    fireEvent.change(fields[1], {target: {value: '0.5'}})
    fireEvent.blur(fields[1])

    expect(fields[1].value).toBe('0.5')

    expect(setValueMock).toHaveBeenCalledWith(durationToTime({
      hours: 1,
      minutes: 0.5,
    }));
  })

  it('can set a view to a value smaller than 0', async ()=>{

    const Comp = () => {

      const [value, setValue] = useState<number | undefined>(60)

      return <DurationFieldsContainer
        {...{
          views: ['days', 'minutes'],
          value: value,
          setValue,
          DurationFieldProps: {
            inputProps: {
              title: 'fieldTitle',
            },
          },
        }}
      />
    }

    const utils = render(
      <Comp/>
    );

    const fields: any[] = utils.getAllByTitle('fieldTitle')

    fireEvent.change(fields[0], {target: {value: '0.5'}})

    expect(fields[0].value).toBe('0.5')

    fireEvent.blur(fields[0])

    expect(fields[0].value).toBe('')
    expect(fields[1].value).toBe('721')
  })


  it('respects min value', async ()=>{

    const Comp = () => {

      const [value, setValue] = useState<number | undefined>(60)

      return <DurationFieldsContainer
          {...{
            views: ['hours', 'minutes'],
            value: value,
            setValue,
            DurationFieldProps: {
              inputProps: {
                title: 'fieldTitle',
                min: "0"
              },
            },
          }}
      />
    }

    const utils = render(
        <Comp/>
    );

    const fields: any[] = utils.getAllByTitle('fieldTitle')

    fireEvent.change(fields[0], {target: {value: '-1'}})

    expect(fields[0].value).toBe('0')

    fireEvent.blur(fields[0])

    expect(fields[0].value).toBe('0')
    expect(fields[1].value).toBe('1')

    fireEvent.change(fields[1], {target: {value: '-1'}});

    expect(fields[1].value).toBe('0');

    fireEvent.blur(fields[1]);
    expect(fields[0].value).toBe('');
    expect(fields[1].value).toBe('');
  })
})
