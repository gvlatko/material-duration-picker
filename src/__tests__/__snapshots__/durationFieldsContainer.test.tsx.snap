// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationFieldsContainer can match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-96gxtk-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
        >
          Hours
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            title="fieldTitle"
            type="number"
            value="5"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1z7n62"
            >
              <span>
                Hours
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
        >
          Minutes
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            title="fieldTitle"
            type="number"
            value="2"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1z7n62"
            >
              <span>
                Minutes
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
